include/zephyr/sys/time_units.h

```diff
@@ -111,70 +111,128 @@ static TIME_CONSTEXPR inline int sys_clock_hw_cycles_per_sec(void)
  *    round_up  - Return the ceiling of the resulting fraction
  *    round_off - Return the nearest value to the resulting fraction
  *                (pass both round_up/off as false to get "round_down")
+ *
+ * All of this must be implemented as expressions so that, when constant,
+ * the results may be used to initialize global variables.
  */
-static TIME_CONSTEXPR ALWAYS_INLINE uint64_t z_tmcvt(uint64_t t, uint32_t from_hz,
-						  uint32_t to_hz, bool const_hz,
-						  bool result32, bool round_up,
-						  bool round_off)
-{
-	bool mul_ratio = const_hz &&
-		(to_hz > from_hz) && ((to_hz % from_hz) == 0U);
-	bool div_ratio = const_hz &&
-		(from_hz > to_hz) && ((from_hz % to_hz) == 0U);
-
-	if (from_hz == to_hz) {
-		return result32 ? ((uint32_t)t) : t;
-	}
-
-	uint64_t off = 0;
-
-	if (!mul_ratio) {
-		uint32_t rdivisor = div_ratio ? (from_hz / to_hz) : from_hz;
-
-		if (round_up) {
-			off = rdivisor - 1U;
-		}
-		if (round_off) {
-			off = rdivisor / 2U;
-		}
-	}
-
-	/* Select (at build time!) between three different expressions for
-	 * the same mathematical relationship, each expressed with and
-	 * without truncation to 32 bits (I couldn't find a way to make
-	 * the compiler correctly guess at the 32 bit result otherwise).
-	 */
-	if (div_ratio) {
-		t += off;
-		if (result32 && (t < BIT64(32))) {
-			return ((uint32_t)t) / (from_hz / to_hz);
-		} else {
-			return t / ((uint64_t)from_hz / to_hz);
-		}
-	} else if (mul_ratio) {
-		if (result32) {
-			return ((uint32_t)t) * (to_hz / from_hz);
-		} else {
-			return t * ((uint64_t)to_hz / from_hz);
-		}
-	} else {
-		if (result32) {
-			return (uint32_t)((t * to_hz + off) / from_hz);
-		} else if (const_hz && Z_TMCVT_USE_FAST_ALGO(from_hz, to_hz)) {
-			/* Faster algorithm but source is first multiplied by target frequency
-			 * and it can overflow even though final result would not overflow.
-			 * Kconfig option shall prevent use of this algorithm when there is a
-			 * risk of overflow.
-			 */
-			return ((t * to_hz + off) / from_hz);
-		} else {
-			/* Slower algorithm but input is first divided before being multiplied
-			 * which prevents overflow of intermediate value.
-			 */
-			return (t / from_hz) * to_hz + ((t % from_hz) * to_hz + off) / from_hz;
-		}
-	}
-}
+
+/* true if the conversion is the identity */
+#define z_tmcvt_identity(__from_hz, __to_hz) \
+	((__to_hz) == (__from_hz))
+
+/* true if the conversion requires a simple integer multiply */
+#define z_tmcvt_mul_ratio(__from_hz, __to_hz) \
+	((__to_hz) > (__from_hz) && (__to_hz) % (__from_hz) == 0U)
+
+/* true if the conversion requires a simple integer division */
+#define z_tmcvt_div_ratio(__from_hz, __to_hz) \
+	((__from_hz) > (__to_hz) && (uint64_t) (__from_hz) % (__to_hz) == 0U)
```

Comment:
> Out of interest: Why is the (uint64_t) cast required? AFAICS no risk of overflowing here.

Good catch. I've removed that.

> I think here and in other instances below some casts do just the same as automatic C type casting. This limits the general usability of the macros and maybe the compiler's leeway for optimization in case other input types will be used?

Agreed. I've removed some additional casts related to the from_hz and to_hz values but left all of the casts for the input value to uint64_t -- that ensures that computations will be wide enough to avoid overflows while still allowing the compiler to do whatever optimizations it can manage with the input and output frequency values.

> Actually the mul/div macros do the same and could be one (switching inputs). The naming of the macro could be improved then - some infix like "is_integer_divisible" or "is_multiple_of" would better capture the actual intention IMO.

Yup, these names came from the original source code, but as they had `bool` type, it was easier to see what they did. I've changed those to `z_tmcvt_is_identity`, `z_tmcvt_is_int_mul` and `z_tmcvt_is_int_div` to make it clear that they are predicates.

> Can be done in a separate commit such that verifying the correctness of the steps remains easy.

The existing tests have been very good at catching implementation errors; I'm going to squash these updates into the original patch as reading the differences is actually harder than just looking at the new code.